File: kotlin/src/main/kotlin/com/edgeveda/sdk/CameraUtils.kt
/**
 * Camera Utilities for Edge Veda Kotlin SDK
 *
 * Pixel-format converters matching Flutter's CameraUtils:
 * - BGRA → RGB (non-standard sources)
 * - YUV420 → RGB (Android camera ImageFormat.YUV_420_888, BT.601 coefficients)
 * - Nearest-neighbor resize
 */
package com.edgeveda.sdk

/**
 * Camera pixel-format conversion and image resizing utilities.
 *
 * These helpers convert raw camera frame buffers into the RGB888 format
 * expected by the vision inference pipeline.
 */
object CameraUtils {

    // ── BGRA → RGB ──────────────────────────────────────────────────────

    /**
     * Convert BGRA8888 pixel data to RGB888.
     *
     * Strips the alpha channel and reorders B-G-R-A → R-G-B.
     *
     * @param bgra Raw BGRA pixel data (width × height × 4 bytes).
     * @param width Frame width in pixels.
     * @param height Frame height in pixels.
     * @return RGB888 data (width × height × 3 bytes).
     * @throws IllegalArgumentException if buffer size is wrong.
     */
    fun convertBgraToRgb(bgra: ByteArray, width: Int, height: Int): ByteArray {
        val expectedSize = width * height * 4
        require(bgra.size >= expectedSize) {
            "BGRA buffer too small: expected $expectedSize bytes, got ${bgra.size}"
        }

        val pixelCount = width * height
        val rgb = ByteArray(pixelCount * 3)

        for (i in 0 until pixelCount) {
            val srcOff = i * 4
            val dstOff = i * 3
            rgb[dstOff]     = bgra[srcOff + 2] // R (from BGRA position 2)
            rgb[dstOff + 1] = bgra[srcOff + 1] // G (from BGRA position 1)
            rgb[dstOff + 2] = bgra[srcOff]     // B (from BGRA position 0)
        }

        return rgb
    }

    // ── YUV420 → RGB (BT.601) ───────────────────────────────────────────

    /**
     * Convert YUV420 planar data to RGB888 using BT.601 coefficients.
     *
     * This matches the conversion used by Flutter's `CameraUtils` for Android
     * camera frames (`ImageFormat.YUV_420_888`).
     *
     * BT.601 coefficients:
     * ```
     * R = Y + 1.402  × (V - 128)
     * G = Y - 0.3441 × (U - 128) - 0.7141 × (V - 128)
     * B = Y + 1.772  × (U - 128)
     * ```
     *
     * @param yPlane Luminance plane (width × height bytes).
     * @param uPlane Chrominance U plane (width/2 × height/2 bytes).
     * @param vPlane Chrominance V plane (width/2 × height/2 bytes).
     * @param width Frame width in pixels (must be even).
     * @param height Frame height in pixels (must be even).
     * @return RGB888 data (width × height × 3 bytes).
     * @throws IllegalArgumentException if buffer sizes are wrong.
     */
    fun convertYuv420ToRgb(
        yPlane: ByteArray,
        uPlane: ByteArray,
        vPlane: ByteArray,
        width: Int,
        height: Int
    ): ByteArray {
        val expectedY = width * height
        val expectedUV = (width / 2) * (height / 2)

        require(yPlane.size >= expectedY) {
            "Y plane too small: expected $expectedY, got ${yPlane.size}"
        }
        require(uPlane.size >= expectedUV) {
            "U plane too small: expected $expectedUV, got ${uPlane.size}"
        }
        require(vPlane.size >= expectedUV) {
            "V plane too small: expected $expectedUV, got ${vPlane.size}"
        }

        val pixelCount = width * height
        val rgb = ByteArray(pixelCount * 3)
        val uvWidth = width / 2

        for (row in 0 until height) {
            for (col in 0 until width) {
                val yIdx = row * width + col
                val uvIdx = (row / 2) * uvWidth + (col / 2)

                val yVal = (yPlane[yIdx].toInt() and 0xFF).toFloat()
                val uVal = (uPlane[uvIdx].toInt() and 0xFF).toFloat() - 128f
                val vVal = (vPlane[uvIdx].toInt() and 0xFF).toFloat() - 128f

                val r = yVal + 1.402f * vVal
                val g = yVal - 0.3441f * uVal - 0.7141f * vVal
                val b = yVal + 1.772f * uVal

                val dstOff = yIdx * 3
                rgb[dstOff]     = clampToByte(r)
                rgb[dstOff + 1] = clampToByte(g)
                rgb[dstOff + 2] = clampToByte(b)
            }
        }

        return rgb
    }

    // ── Nearest-Neighbor Resize ─────────────────────────────────────────

    /**
     * Resize RGB888 image data using nearest-neighbor interpolation.
     *
     * A fast, low-quality resize suitable for preparing camera frames
     * for vision model input where sub-pixel accuracy is not critical.
     *
     * @param rgb Source RGB888 data (srcWidth × srcHeight × 3 bytes).
     * @param srcWidth Source width in pixels.
     * @param srcHeight Source height in pixels.
     * @param dstWidth Destination width in pixels.
     * @param dstHeight Destination height in pixels.
     * @return Resized RGB888 data (dstWidth × dstHeight × 3 bytes).
     * @throws IllegalArgumentException if buffer size is wrong or dimensions invalid.
     */
    fun resizeRgb(
        rgb: ByteArray,
        srcWidth: Int,
        srcHeight: Int,
        dstWidth: Int,
        dstHeight: Int
    ): ByteArray {
        val expectedSize = srcWidth * srcHeight * 3
        require(rgb.size >= expectedSize) {
            "RGB buffer too small: expected $expectedSize bytes, got ${rgb.size}"
        }
        require(dstWidth > 0 && dstHeight > 0) {
            "Destination dimensions must be positive: ${dstWidth}×${dstHeight}"
        }

        // Short-circuit: no resize needed
        if (srcWidth == dstWidth && srcHeight == dstHeight) {
            return rgb.copyOf()
        }

        val result = ByteArray(dstWidth * dstHeight * 3)
        val xRatio = srcWidth.toFloat() / dstWidth.toFloat()
        val yRatio = srcHeight.toFloat() / dstHeight.toFloat()

        for (dstY in 0 until dstHeight) {
            val srcY = minOf((dstY * yRatio).toInt(), srcHeight - 1)
            for (dstX in 0 until dstWidth) {
                val srcX = minOf((dstX * xRatio).toInt(), srcWidth - 1)

                val srcOff = (srcY * srcWidth + srcX) * 3
                val dstOff = (dstY * dstWidth + dstX) * 3

                result[dstOff]     = rgb[srcOff]
                result[dstOff + 1] = rgb[srcOff + 1]
                result[dstOff + 2] = rgb[srcOff + 2]
            }
        }

        return result
    }

    // ── Helpers ─────────────────────────────────────────────────────────

    /** Clamp a float to the 0–255 byte range. */
    private fun clampToByte(value: Float): Byte {
        return value.coerceIn(0f, 255f).toInt().toByte()
    }
}

File: web/src/CameraUtils.ts
/**
 * CameraUtils — Pixel-format conversion utilities for the Web SDK.
 *
 * Provides BGRA→RGB, YUV420→RGB (BT.601), and nearest-neighbor resize.
 * All methods operate on Uint8Array buffers and are pure functions
 * (no Canvas dependency) so they work in Web Workers too.
 */

/** Clamp a number to [0, 255] and truncate to integer. */
function clampToByte(v: number): number {
  return v < 0 ? 0 : v > 255 ? 255 : v | 0;
}

export class CameraUtils {
  /**
   * Convert a BGRA buffer to packed RGB.
   *
   * @param bgra  Source buffer — length must equal `width * height * 4`.
   * @param width  Image width in pixels.
   * @param height Image height in pixels.
   * @returns Uint8Array of length `width * height * 3`.
   */
  static convertBgraToRgb(
    bgra: Uint8Array,
    width: number,
    height: number,
  ): Uint8Array {
    const pixelCount = width * height;
    const expectedLen = pixelCount * 4;
    if (bgra.length !== expectedLen) {
      throw new Error(
        `CameraUtils.convertBgraToRgb: expected ${expectedLen} bytes, got ${bgra.length}`,
      );
    }

    const rgb = new Uint8Array(pixelCount * 3);
    let srcIdx = 0;
    let dstIdx = 0;

    for (let i = 0; i < pixelCount; i++) {
      // BGRA → RGB: swap B and R, drop A
      rgb[dstIdx] = bgra[srcIdx + 2]; // R
      rgb[dstIdx + 1] = bgra[srcIdx + 1]; // G
      rgb[dstIdx + 2] = bgra[srcIdx]; // B
      srcIdx += 4;
      dstIdx += 3;
    }

    return rgb;
  }

  /**
   * Convert planar YUV 420 to packed RGB using BT.601 coefficients.
   *
   * Chroma planes are half-resolution in both dimensions.
   * Coefficients:
   *   R = Y + 1.402  × (V − 128)
   *   G = Y − 0.3441 × (U − 128) − 0.7141 × (V − 128)
   *   B = Y + 1.772  × (U − 128)
   *
   * @param yPlane  Luma plane — length `width * height`.
   * @param uPlane  Cb plane  — length `(width >> 1) * (height >> 1)`.
   * @param vPlane  Cr plane  — length `(width >> 1) * (height >> 1)`.
   * @param width   Image width (must be even).
   * @param height  Image height (must be even).
   * @returns Uint8Array of length `width * height * 3`.
   */
  static convertYuv420ToRgb(
    yPlane: Uint8Array,
    uPlane: Uint8Array,
    vPlane: Uint8Array,
    width: number,
    height: number,
  ): Uint8Array {
    const pixelCount = width * height;
    const uvWidth = width >> 1;
    const uvHeight = height >> 1;

    if (yPlane.length !== pixelCount) {
      throw new Error(
        `CameraUtils.convertYuv420ToRgb: yPlane length ${yPlane.length} != ${pixelCount}`,
      );
    }
    if (uPlane.length !== uvWidth * uvHeight) {
      throw new Error(
        `CameraUtils.convertYuv420ToRgb: uPlane length ${uPlane.length} != ${uvWidth * uvHeight}`,
      );
    }
    if (vPlane.length !== uvWidth * uvHeight) {
      throw new Error(
        `CameraUtils.convertYuv420ToRgb: vPlane length ${vPlane.length} != ${uvWidth * uvHeight}`,
      );
    }

    const rgb = new Uint8Array(pixelCount * 3);
    let dstIdx = 0;

    for (let row = 0; row < height; row++) {
      const uvRow = row >> 1;
      for (let col = 0; col < width; col++) {
        const uvCol = col >> 1;

        const y = yPlane[row * width + col];
        const u = uPlane[uvRow * uvWidth + uvCol];
        const v = vPlane[uvRow * uvWidth + uvCol];

        const uShifted = u - 128;
        const vShifted = v - 128;

        rgb[dstIdx] = clampToByte(y + 1.402 * vShifted);
        rgb[dstIdx + 1] = clampToByte(y - 0.3441 * uShifted - 0.7141 * vShifted);
        rgb[dstIdx + 2] = clampToByte(y + 1.772 * uShifted);
        dstIdx += 3;
      }
    }

    return rgb;
  }

  /**
   * Nearest-neighbour resize of a packed RGB buffer.
   *
   * @param rgb   Source buffer — length `srcWidth * srcHeight * 3`.
   * @param srcWidth  Source width.
   * @param srcHeight Source height.
   * @param dstWidth  Target width.
   * @param dstHeight Target height.
   * @returns Uint8Array of length `dstWidth * dstHeight * 3`.
   */
  static resizeRgb(
    rgb: Uint8Array,
    srcWidth: number,
    srcHeight: number,
    dstWidth: number,
    dstHeight: number,
  ): Uint8Array {
    const expectedLen = srcWidth * srcHeight * 3;
    if (rgb.length !== expectedLen) {
      throw new Error(
        `CameraUtils.resizeRgb: expected ${expectedLen} bytes, got ${rgb.length}`,
      );
    }

    const out = new Uint8Array(dstWidth * dstHeight * 3);
    const xRatio = srcWidth / dstWidth;
    const yRatio = srcHeight / dstHeight;

    let dstIdx = 0;

    for (let dy = 0; dy < dstHeight; dy++) {
      const srcY = (dy * yRatio) | 0;
      for (let dx = 0; dx < dstWidth; dx++) {
        const srcX = (dx * xRatio) | 0;
        const srcIdx = (srcY * srcWidth + srcX) * 3;
        out[dstIdx] = rgb[srcIdx];
        out[dstIdx + 1] = rgb[srcIdx + 1];
        out[dstIdx + 2] = rgb[srcIdx + 2];
        dstIdx += 3;
      }
    }

    return out;
  }
}